File: src/scenes/lorenz_attractor.rs
use crate::engine::{GlobalContext, Scene};
use crate::utils::RK4::rk4;
use sdl2::sys::KeyCode;
use sdl2::{event::Event, keyboard::Keycode};
use sdl2::{render::Canvas, video::Window};

pub struct LorenzAttractor {
    sigma: f32,
    beta: f32,
    rho: f32,
    solutions: Option<(Vec<f32>, Vec<Vec<f32>>)>,
    current_index: usize,
    done: bool,
    camera_rotation: (f32, f32),
    is_mouse_down: bool,
    zoom: f32,
}

impl LorenzAttractor {
    pub fn new() -> Self {
        let mut lorenz_attractor = LorenzAttractor {
            sigma: 10.0,
            beta: 2.667,
            rho: 28.0,
            solutions: None,
            current_index: 0,
            done: false,
            camera_rotation: (0.0, 0.0),
            is_mouse_down: false,
            zoom: 1.0,
        };
        lorenz_attractor.solve();
        lorenz_attractor
    }

    fn lorenz(&self, _t: f32, state: &[f32; 3]) -> [f32; 3] {
        let (x, y, z) = (state[0], state[1], state[2]);
        let x_dot = self.sigma * (y - x);
        let y_dot = x * (self.rho - z) - y;
        let z_dot = x * y - self.beta * z;
        [x_dot, y_dot, z_dot]
    }

    pub fn solve(&mut self) {
        let r0 = [0.0, 1.0, 1.05];
        let (a, b) = (0.0, 50.0);
        let n: u32 = 10000;
        let solutions = rk4(a, b, r0, |t, state| self.lorenz(t, state), n);
        for value in &solutions.1 {
            print!("Solution x:{}, y:{}, z:{} \n", value[0], value[1], value[2]);
        }
        self.solutions = Some(solutions);
    }

    fn rotate_3d(&self, point: (f32, f32, f32)) -> (f32, f32, f32) {
        let (x, y, z) = point;
        let (theta_x, theta_y) = self.camera_rotation;

        let (x1, y1, z1) = (
            x,
            y * theta_x.cos() - z * theta_x.sin(),
            y * theta_x.sin() + z * theta_x.cos(),
        );

        let (x2, y2, z2) = (
            x1 * theta_y.cos() + z1 * theta_y.sin(),
            y1,
            -x1 * theta_y.sin() + z1 * theta_y.cos(),
        );

        (x2, y2, z2)
    }

    fn project(&self, point: (f32, f32, f32), width: u32, height: u32) -> (i32, i32) {
        let (x, y, z) = point;

        let d = 50.0;
        let x_screen = x / (z + d) * 200.0 * self.zoom + width as f32 / 2.0;
        let y_screen = y / (z + d) * 200.0 * self.zoom + height as f32 / 2.0;

        (x_screen as i32, y_screen as i32)
    }
}

impl Scene for LorenzAttractor {
    fn update(&mut self, ctx: &mut GlobalContext, _dt: f32) {
        if self.solutions.is_none() {
            self.solve();
        }

        if let Some((_, ref sol)) = self.solutions {
            if (self.current_index + 1 + (ctx.simulation_speed * 10.) as usize) < sol.len() {
                self.current_index += 1 + (ctx.simulation_speed * 10.) as usize;
            } else {
                self.current_index = sol.len() - 1
            }
        }
    }

    fn render(&mut self, _ctx: &GlobalContext, canvas: &mut Canvas<Window>) {
        if let Some((_, ref solutions)) = self.solutions {
            canvas.set_draw_color(sdl2::pixels::Color::RGB(255, 255, 255));

            let (width, height) = canvas.output_size().unwrap_or((800, 600));
            let mut last_projected = None;

            let total_points = solutions.len();

            for i in 1..self.current_index {
                let point = (solutions[i][0], solutions[i][1], solutions[i][2]);

                let rotated = self.rotate_3d(point);
                let projected = self.project(rotated, width, height);
                let t = i as f32 / total_points as f32;
                let r = (255.0 * (1.0 - t)) as u8;
                let g = (255.0 * t) as u8;

                if let Some(last) = last_projected {
                    canvas.set_draw_color(sdl2::pixels::Color::RGB(r, g, 0));
                    let _ = canvas.draw_line(last, projected);
                }
                last_projected = Some(projected);
            }
        }
    }

    fn handle_event(&mut self, ctx: &mut GlobalContext, event: &Event) {
        match event {
            Event::MouseButtonDown { mouse_btn, .. } => {
                if *mouse_btn == sdl2::mouse::MouseButton::Left {
                    self.is_mouse_down = true;
                }
            }
            Event::MouseButtonUp { mouse_btn, .. } => {
                if *mouse_btn == sdl2::mouse::MouseButton::Left {
                    self.is_mouse_down = false;
                }
            }
            Event::MouseMotion { xrel, yrel, .. } => {
                if self.is_mouse_down {
                    self.camera_rotation.0 += *yrel as f32 * 0.01;
                    self.camera_rotation.1 += *xrel as f32 * 0.01;
                }
            }
            Event::KeyDown {
                keycode: Some(k),
                ..
            } => {
                match k {
                    Keycode::Escape => {
                        self.done = true;
                    },
                    Keycode::R => {
                        self.current_index = 0;
                    }
                    Keycode::Left => {
                        ctx.simulation_speed -= 0.1;
                    }
                    Keycode::Right => {
                        ctx.simulation_speed += 0.1;
                    }
                    _ => {}
                }
            }
            Event::MouseWheel { y, .. } => {
                if *y > 0 {
                    self.zoom *= 1.1; 
                } else if *y < 0 {
                    self.zoom *= 0.9; 
                }
            }
            _ => {}
        }
    }

    fn is_done(&self) -> bool {
        self.done
    }
}

File: c++/scenes/lorenz_attractor.cpp
#include "../include/lorenz_attractor.h"
#include "../include/utils/rk4.h"
#include <SDL2/SDL.h>
#include <cmath>

LorenzAttractor::LorenzAttractor()
    : render_mode(RenderMode::Points), // Modo predeterminado: puntos blancos
      sigma(10.0f), beta(2.667f), rho(28.0f),
      solutions(std::nullopt), current_index(0), done(false),
      camera_rotation{0.0f, 0.0f}, is_mouse_down(false), zoom(1.0f)
{
    solve();
}

void LorenzAttractor::solve()
{
    auto lorenz = [this](float t, const std::array<float, 3> &state)
    {
        return this->lorenz(t, state);
    };

    auto r0 = std::array<float, 3>{0.0f, 1.0f, 1.05f};
    auto [t_values, w] = rk4(0.0f, 50.0f, r0, lorenz, 10000);
    solutions = {t_values, w};
    for (auto &sol : w)
    {
        printf("x: %f, y: %f, z: %f", sol[0], sol[1], sol[2]);
    }
}
void LorenzAttractor::setRenderMode(RenderMode mode)
{
    render_mode = mode;
}

std::array<float, 3> LorenzAttractor::lorenz(float t, const std::array<float, 3> &state)
{
    float x = state[0], y = state[1], z = state[2];
    return {sigma * (y - x), x * (rho - z) - y, x * y - beta * z};
}

std::pair<int, int> LorenzAttractor::project(const std::array<float, 3> &point, int width, int height)
{
    float d = 50.0f;
    float x_screen = point[0] / (point[2] + d) * 200.0f * zoom + width / 2.0f;
    float y_screen = point[1] / (point[2] + d) * 200.0f * zoom + height / 2.0f;
    return {static_cast<int>(x_screen), static_cast<int>(y_screen)};
}

std::array<float, 3> LorenzAttractor::rotate3D(const std::array<float, 3> &point)
{
    float x = point[0], y = point[1], z = point[2];
    float theta_x = camera_rotation.first;
    float theta_y = camera_rotation.second;

    float y1 = y * cos(theta_x) - z * sin(theta_x);
    float z1 = y * sin(theta_x) + z * cos(theta_x);

    float x2 = x * cos(theta_y) + z1 * sin(theta_y);
    float z2 = -x * sin(theta_y) + z1 * cos(theta_y);

    return {x2, y1, z2};
}

void LorenzAttractor::handleEvent(GlobalContext &ctx, const SDL_Event &event)
{
    switch (event.type)
    {
    case SDL_MOUSEBUTTONDOWN:
        if (event.button.button == SDL_BUTTON_LEFT)
        {
            is_mouse_down = true;
        }
        break;

    case SDL_MOUSEBUTTONUP:
        if (event.button.button == SDL_BUTTON_LEFT)
        {
            is_mouse_down = false;
        }
        break;

    case SDL_MOUSEMOTION:
        if (is_mouse_down)
        {
            camera_rotation.first += event.motion.yrel * 0.01f;  // Rotación en X
            camera_rotation.second += event.motion.xrel * 0.01f; // Rotación en Y
        }
        break;
    case SDL_KEYDOWN:
        switch (event.key.keysym.sym)
        {
        case SDLK_ESCAPE: // Escape
            done = true;
            break;
        case SDLK_r: // Reset
            current_index = 0;
            break;
        case SDLK_LEFT: // Disminuir velocidad
            ctx.simulation_speed = std::max(0.1f, ctx.simulation_speed - 0.1f);
            break;
        case SDLK_RIGHT: // Aumentar velocidad
            ctx.simulation_speed += 0.1f;
            break;
        case SDLK_m: // Cambiar modo de renderizado
            if (render_mode == RenderMode::Points)
            {
                render_mode = RenderMode::GradientLines;
            }
            else
            {
                render_mode = RenderMode::Points;
            }
            break;
        default:
            break;
        }
        break;

    case SDL_MOUSEWHEEL:
        if (event.wheel.y > 0)
        { // Zoom in
            zoom *= 1.1f;
        }
        else if (event.wheel.y < 0)
        { // Zoom out
            zoom *= 0.9f;
        }
        break;

    default:
        break;
    }
}
void LorenzAttractor::update(GlobalContext &ctx, float dt)
{
    if (solutions)
    {
        const auto &sol = solutions->second;
        current_index = std::min(current_index + static_cast<int>(ctx.simulation_speed * 10.0f), static_cast<int>(sol.size()) - 1);
    }
}

void LorenzAttractor::render(GlobalContext &ctx, SDL_Renderer *renderer)
{
    if (!solutions)
        return;

    const auto &sol = solutions->second;
    size_t total_points = sol.size();

    std::pair<int, int> last_projected;

    for (size_t i = 1; i < current_index; ++i)
    {
        auto point = sol[i];
        auto rotated = rotate3D(point);
        auto projected = project(rotated, ctx.screen_width, ctx.screen_height);

        if (render_mode == RenderMode::Points)
        {
            SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255); // Blanco
            SDL_RenderDrawPoint(renderer, projected.first, projected.second);
        }
        else if (render_mode == RenderMode::GradientLines)
        {
            float t = static_cast<float>(i) / total_points;
            Uint8 r = static_cast<Uint8>(255.0f * (1.0f - t));
            Uint8 g = static_cast<Uint8>(255.0f * t);

            if (i > 1)
            {
                SDL_SetRenderDrawColor(renderer, r, g, 0, 255);
                SDL_RenderDrawLine(renderer, last_projected.first, last_projected.second,
                                   projected.first, projected.second);
            }
        }

        last_projected = projected;
    }
}

bool LorenzAttractor::isDone() const
{
    return done;
}
